### PR TITLE
Add tests for a ShoppingCart functionality

### PR DESCRIPTION
**Pull request by nikitaabo**

### CODE REVIEW

**Comment by nikitaabo** [comment]
Screenshot 2024-10-05 at 03 25 21

File: src/test/java/com/example/spring/controller/ShoppingCartControllerTest.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+import org.springframework.test.web.servlet.MvcResult;
+import org.springframework.test.web.servlet.setup.MockMvcBuilders;
+import org.springframework.web.context.WebApplicationContext;
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+@Sql(
+        scripts = {
+                "classpath:database/shopping.carts/clear-cart.sql",
+                "classpath:database/cart.items/clear-cart-items.sql",
+                "classpath:database/books/clear-books.sql",
```
Comment (liliia-ponomarenko) [inline src/test/java/com/example/spring/controller/ShoppingCartControllerTest.java]: Why do you need to execute these scripts before and after methods? Can you do it only after each test method?

File: src/test/java/com/example/spring/controller/ShoppingCartControllerTest.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    @BeforeAll
+    static void beforeAll(@Autowired WebApplicationContext applicationContext) {
+        mockMvc = MockMvcBuilders
+                .webAppContextSetup(applicationContext)
+                .apply(springSecurity())
+                .build();
+    }
+
+    @Test
+    @DisplayName("Get the shopping cart for a user")
```
Comment (liliia-ponomarenko) [inline src/test/java/com/example/spring/controller/ShoppingCartControllerTest.java:67]: Better use  `WithMockUser(username = ... , roles = ...)` instead jwt util

File: src/test/java/com/example/spring/controller/ShoppingCartControllerTest.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        MvcResult result = mockMvc.perform(get("/cart")
+                        .header("Authorization", "Bearer " + jwtToken)
+                        .contentType(MediaType.APPLICATION_JSON_VALUE))
+                .andExpect(status().isOk())
+                .andReturn();
+
+        // Then
+        ShoppingCartDto cartDto = objectMapper.readValue(
+                result.getResponse().getContentAsString(), ShoppingCartDto.class);
+        assertNotNull(cartDto);
```
Comment (liliia-ponomarenko) [inline src/test/java/com/example/spring/controller/ShoppingCartControllerTest.java:80]: You also should check the content of the response ;)

File: src/test/java/com/example/spring/controller/ShoppingCartControllerTest.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    }
+
+    private String generateJwtTokenForTestUser() {
+        return jwtUtil.generateToken("testUser@test.com");
+    }
+
+    @Test
+    @WithMockUser(username = "admin", roles = {"ADMIN"})
```
Comment (liliia-ponomarenko) [inline src/test/java/com/example/spring/controller/ShoppingCartControllerTest.java:87]: Like here ) Why do you use different types of authorization in other test cases?

File: src/test/java/com/example/spring/controller/ShoppingCartControllerTest.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+        // Given
+        Long cartItemId = 1L;
+
+        // When
+        mockMvc.perform(delete("/cart/items/{id}", cartItemId)
+                        .contentType(MediaType.APPLICATION_JSON_VALUE))
+                .andExpect(status().isNoContent());
+    }
+}
+
```
Comment (liliia-ponomarenko) [inline src/test/java/com/example/spring/controller/ShoppingCartControllerTest.java:98]: Only one empty line at the end of the file

File: src/test/java/com/example/spring/controller/ShoppingCartControllerTest.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+        scripts = {
+                "classpath:database/users-roles/clear-users-roles.sql",
+                "classpath:database/roles/clear-role.sql",
+                "classpath:database/cart.items/clear-cart-items.sql",
+                "classpath:database/shopping.carts/clear-cart.sql",
+                "classpath:database/books/clear-books.sql"
+        },
+        executionPhase = Sql.ExecutionPhase.AFTER_TEST_METHOD
+)
+public class ShoppingCartControllerTest {
```
Comment (liliia-ponomarenko) [inline src/test/java/com/example/spring/controller/ShoppingCartControllerTest.java:52]: I see that you have more than two controllers in this class. Let's test others ;)

File: src/test/java/com/example/spring/service/ShoppingCartServiceTest.java [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+        cartItem.setBook(book);
+        cartItem.setShoppingCart(shoppingCart);
+
+        when(bookRepository.findById(requestCartItemDto.bookId())).thenReturn(Optional.of(book));
+        when(shoppingCartRepository.findByUserEmail(user.getEmail())).thenReturn(shoppingCart);
+        when(cartItemRepository.existsByBook(book)).thenReturn(true);
+        when(cartItemRepository.findByBook(book)).thenReturn(List.of(cartItem));
+
+        // When & Then
+        assertThrows(DataProcessingException.class,
```
Comment (liliia-ponomarenko) [inline src/test/java/com/example/spring/service/ShoppingCartServiceTest.java]: Don't test system exceptions! 
I guess you don't have any logic for handling this exception

File: src/test/resources/database/cart.items/add-cart-items.sql [hunk 8]
```diff
@@ -0,0 +1,2 @@
+INSERT INTO cart_items (id, shopping_cart_id, book_id, quantity) VALUES (1, 1, 1, 2);
+
```
Comment (liliia-ponomarenko) [inline src/test/resources/database/cart.items/add-cart-items.sql]: Only one empty line at the end of the file

File: src/test/resources/database/roles/add-role.sql [hunk 9]
```diff
@@ -0,0 +1,2 @@
+DELETE FROM roles WHERE name = 'ROLE_USER';
```
Comment (liliia-ponomarenko) [inline src/test/resources/database/roles/add-role.sql]: Don't use deleting in the insert script. It's not really clear

File: src/test/resources/database/shopping.carts/add-cart.sql [hunk 10]
```diff
@@ -0,0 +1,7 @@
+DELETE from users;
```
Comment (liliia-ponomarenko) [inline src/test/resources/database/shopping.carts/add-cart.sql]: the same

File: src/test/resources/database/shopping.carts/add-cart.sql [hunk 11]
```diff
@@ -0,0 +1,7 @@
...
+
+INSERT INTO users (id, email, password, first_name, last_name, shipping_address, is_deleted)
+VALUES (1, 'testUser@test.com', 'password123', 'John', 'Doe', '123 Test St', false);
```
Comment (liliia-ponomarenko) [inline src/test/resources/database/shopping.carts/add-cart.sql]: Should you encode a password?

File: src/test/resources/database/users-roles/add-user-role.sql [hunk 12]
```diff
@@ -0,0 +1,3 @@
+INSERT INTO users_roles (user_id, role_id)
+VALUES ((SELECT id FROM users WHERE email='testUser@test.com'),
+        (SELECT id FROM roles WHERE name='ROLE_USER'));
```
Comment (liliia-ponomarenko) [inline src/test/resources/database/users-roles/add-user-role.sql]: you can add rows with ids and avoid select statements here

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution ;)

File: src/test/java/com/example/spring/service/ShoppingCartServiceTest.java (hunk 7 above)
Comment (JJJazl) [reply src/test/java/com/example/spring/service/ShoppingCartServiceTest.java]: Please, check this review comment about system exceptions

File: src/main/java/com/example/spring/controller/ShoppingCartController.java [hunk 13]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (JJJazl) [inline src/main/java/com/example/spring/controller/ShoppingCartController.java]: Remove empty line

File: src/test/java/com/example/spring/controller/ShoppingCartControllerTest.java [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+        scripts = {
+                "classpath:database/roles/clear-role.sql",
+                "classpath:database/users/add-user.sql",
+                "classpath:database/shopping.carts/add-cart.sql",
+                "classpath:database/roles/add-role.sql",
+                "classpath:database/users-roles/add-user-role.sql",
+                "classpath:database/books/add-book.sql",
+                "classpath:database/cart.items/add-cart-items.sql"
+        },
+        executionPhase = Sql.ExecutionPhase.BEFORE_TEST_METHOD
```
Comment (Elena-Bruyako) [inline src/test/java/com/example/spring/controller/ShoppingCartControllerTest.java]: BEFORE_TEST_METHOD by default you can remove executionPhase

File: src/test/java/com/example/spring/controller/ShoppingCartControllerTest.java (hunk 6 above)
Comment (Elena-Bruyako) [reply src/test/java/com/example/spring/controller/ShoppingCartControllerTest.java:52]: comment not fixed

File: src/test/java/com/example/spring/service/ShoppingCartServiceTest.java [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+
+        // When
+        ShoppingCartDto result = shoppingCartService.getCart(email);
+
+        // Then
+        assertNotNull(result);
+        assertEquals(shoppingCartDto, result);
+        verify(shoppingCartRepository, times(1)).findByUserEmail(email);
+        verify(shoppingCartMapper, times(1)).toDto(shoppingCart);
+    }
```
Comment (Elena-Bruyako) [inline src/test/java/com/example/spring/service/ShoppingCartServiceTest.java:155]: Add updateCartItem test case

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]

File: src/test/java/com/example/spring/controller/ShoppingCartControllerTest.java (hunk 6 above)
Comment (nikitaabo) [reply src/test/java/com/example/spring/controller/ShoppingCartControllerTest.java:52]: I didn't get it. Where do I have more than two controllers in this class? What have I to do?

File: src/test/java/com/example/spring/service/ShoppingCartServiceTest.java (hunk 15 above)
Comment (ahoienko) [reply src/test/java/com/example/spring/service/ShoppingCartServiceTest.java:155]: Not resolved

**Review by ahoienko — CHANGES_REQUESTED** [review]
Please read comment carefully

**Review by sarakhmen — APPROVED** [review]

**Pull request merged**